flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java

```diff
@@ -76,36 +76,25 @@ public ApplicationReconciler(
     }
 
     @Override
-    protected AvailableUpgradeMode getAvailableUpgradeMode(
+    protected JobUpgrade getJobUpgrade(
             FlinkResourceContext<FlinkDeployment> ctx, Configuration deployConfig)
             throws Exception {
 
         var deployment = ctx.getResource();
         var status = deployment.getStatus();
-        var availableUpgradeMode = super.getAvailableUpgradeMode(ctx, deployConfig);
+        var availableUpgradeMode = super.getJobUpgrade(ctx, deployConfig);
 
         if (availableUpgradeMode.isAvailable() || !availableUpgradeMode.isAllowFallback()) {
             return availableUpgradeMode;
         }
         var flinkService = ctx.getFlinkService();
 
-        boolean lastStateAllowed =
-                deployment.getSpec().getJob().getUpgradeMode() == UpgradeMode.LAST_STATE
-                        || deployConfig.getBoolean(
-                                KubernetesOperatorConfigOptions
-                                        .OPERATOR_JOB_UPGRADE_LAST_STATE_FALLBACK_ENABLED);
-
-        if (lastStateAllowed
-                && HighAvailabilityMode.isHighAvailabilityModeActivated(deployConfig)
+        if (HighAvailabilityMode.isHighAvailabilityModeActivated(deployConfig)
                 && HighAvailabilityMode.isHighAvailabilityModeActivated(ctx.getObserveConfig())
-                && !flinkVersionChanged(
-                        ReconciliationUtils.getDeployedSpec(deployment), deployment.getSpec())) {
-
-            if (flinkService.isHaMetadataAvailable(deployConfig)) {
-                LOG.info(
-                        "Job is not running but HA metadata is available for last state restore, ready for upgrade");
-                return AvailableUpgradeMode.of(UpgradeMode.LAST_STATE);
-            }
+                && flinkService.isHaMetadataAvailable(deployConfig)) {
```

Comment:
Is it safe to remove the Flink version change check here?

Reply:
Yes, the version change scenario is fully handled in the AbstractJobReconciler and cannot reach this point. In those cases we always return Savepoint/Cancel/Unavailable